src/formatting_stack/util/diff.clj

```diff
@@ -0,0 +1,69 @@
+(ns formatting-stack.util.diff
+  (:require
+   [clojure.java.data :as data]
+   [clojure.java.io :as io]
+   [clojure.spec.alpha :as spec]
+   [clojure.string :as string]
+   [formatting-stack.util :refer [rcomp]]
+   [nedap.speced.def :as speced]
+   [nedap.utils.spec.predicates :refer [present-string?]])
+  (:import
+   (difflib DiffUtils)
+   (io.reflectoring.diffparser.api UnifiedDiffParser)
+   (java.io File)
+   (java.util.regex Pattern)))
+
+(spec/def ::start pos-int?)
+(spec/def ::end pos-int?)
+(spec/def ::filename present-string?)
+
+(speced/def-with-doc ::line-numbers
+  "maps of consecutive removals/changes (as `:start` to `:end`) per `:filename`."
+  (spec/coll-of (spec/keys :req-un [::start ::end ::filename])))
+
+(speced/defn ^::line-numbers diff->line-numbers
+  [^string? diff]
```

Comment:
>  `[^string? diff]`

It seems important to spec the first line ,such as:

```
diff --git a/test/unit/formatting_stack/util.clj b/test/unit/formatting_stack/util.clj
```

* left and right hand should refer to existing files
  * guards us against race conditions, or various plausible bugs
* left and right hand should probably have a "normalized" form 
  * relative or absolute filepaths, but not something unspecified
* Can `/dev/null` appear?
  * often does in git diffs

Reply:
> It seems important to spec the first line

I don't see the importance. The `UnifiedDiffParser#parse` method literally parses the diff (and will throw `IllegalStateException` if it runs into an error) which is far more extensive and expressive than a regex could be.

> left and right hand should refer to existing files

> left and right hand should probably have a "normalized" form

Seems overly tight. A diff does not need to be accessible in order be parsable. arbitrary restrictions like these can come as a surprise in the future.

If a consumer requires the path to be accessible, it can specify it itself.